python/taichi/math/_complex.py

```diff
@@ -0,0 +1,202 @@
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def test():
+        >>>     z1 = ti.math.vec2(1, 1)
+        >>>     z2 = ti.math.vec2(0, 1)
+        >>>     ti.math.cdiv(z1, z2)  # [1, -1]
+
+    Returns:
+        :class:`~taichi.math.vec2`: the complex division of `z1 / z2`.
+    """
+    x1, y1 = z1[0], z1[1]
+    x2, y2 = z2[0], z2[1]
+    return vec2(x1 * x2 + y1 * y2, -x1 * y2 + x2 * y1) / dot(z1, z2)
+
+
+@ti.func
+def csqrt(z):
+    """Returns the complex square root of a 2d vector `z`, so that
+    if `w^2=z`, then `w = csqrt(z)`.
+
+    Among the two square roots of `z`, if their real parts are non-zero,
+    the one with positive real part is returned. If both their real parts
+    are zero, the one with non-negative imaginary part is returned.
+
+    Args:
+        z (:class:`~taichi.math.vec2`): The input.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def test():
+        >>>     z = ti.math.vec2(-1, 0)
+        >>>     w = ti.math.csqrt(z)  # [0, 1]
+
+    Returns:
+        :class:`~taichi.math.vec2`: The complex square root.
+    """
+    result = vec2(0.)
+    if any(z):
+        r = ti.sqrt(ti.sqrt(dot(z, z)))
+        a = ti.atan2(z[1], z[0])
+        result = r * vec2(ti.cos(a / 2.0), ti.sin(a / 2.0))
+
+    return result
+
+
+@ti.func
+def cinv(z):
+    """Computes the reciprocal of a complex `z`.
+
+    Args:
+        z (:class:`~taichi.math.vec2`): The input.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def test():
+        >>>     z = ti.math.vec2(1, 1)
+        >>>     w = ti.math.cinv(z)  # [0.5, -0.5]
+
+    Returns:
+        :class:`~taichi.math.vec2`: The reciprocal of `z`.
+    """
+    return cconj(z) / dot(z, z)
+
+
+@ti.func
+def cpow(z, n):
+    """Computes the power of a complex `z`: :math:`z^a`.
+
+    Args:
+        z (:class:`~taichi.math.vec2`): The base.
+        a (float): The exponent.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def test():
+        >>>     z = ti.math.vec2(1, 1)
+        >>>     w = ti.math.cpow(z)  # [-2, 2]
+
+    Returns:
+        :class:`~taichi.math.vec2`: The power :math:`z^a`.
+    """
+    result = vec2(0.)
+    if any(z):
+        r2 = dot(z, z)
+        a = ti.atan2(z[1], z[0]) * n
+        result = ti.pow(r2, n / 2.0) * vec2(ti.cos(a), ti.sin(a))
+
+    return result
+
+
+@ti.func
+def cexp(z):
+    """Returns the complex exponential :math:`e^z`.
+
+    `z` is a 2d vector treated as a complex number.
+
+    Args:
+        z (:class:`~taichi.math.vec2`): The base.
+        a (float): The exponent.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def test():
+        >>>     z = ti.math.vec2(1, 1)
+        >>>     w = ti.math.cexp(z)  # [1.468694, 2.287355]
+
+    Returns:
+        :class:`~taichi.math.vec2`: The power :math:`z^a`
+    """
+    r = ti.exp(z[0])
+    return vec2(r * ti.cos(z[1]), r * ti.sin(z[1]))
+
+
+@ti.func
+def clog(z):
+    """Returns the complex logarithm of `z`.
```

Comment:
Maybe we should point out the range of the result, like `i(-pi, pi] ` or whatever.